Separate consumer/producer settings for Kafka

AVMusorin opened the pull request:
Closes: #58983

**Why does this happen?**

`librdkafka` prints this kind of warning if consumer properties are used for producer instances or producer properties for consumers. https://github.com/ClickHouse/librdkafka/blob/master/src/rdkafka_conf.c#L3940

**New configuration example:**

```xml
<kafka>
        <heartbeat_interval_ms>301</heartbeat_interval_ms>
        <statistics_interval_ms>600</statistics_interval_ms>
       
       <kafka_topic>
            <name>football</name>
            <request_timeout_ms>6000</request_timeout_ms>
        </kafka_topic>
       
        <producer>
          <compression_codec>gzip</compression_codec>
        </producer>

       <consumer>
          <enable_auto_commit>true</enable_auto_commit>
        </consumer>
</kafka>
```

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce separate consumer/producer tags for the Kafka configuration. This avoids warnings from librdkafka that consumer properties were specified for producer instances and vice versa (e.g. `Configuration property session.timeout.ms is a consumer property and will be ignored by this producer instance`). Closes: #58983

azat commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -870,6 +888,21 @@ void StorageKafka::updateConfiguration(cppkafka::Configuration & kafka_config)
+        loadConsumerConfig(kafka_config, config, config_prefix, topics);
+}
+
+void StorageKafka::updateProducerConfiguration(cppkafka::Configuration & kafka_config) {
```
Coding style issue

```suggestion
void StorageKafka::updateProducerConfiguration(cppkafka::Configuration & kafka_config)
{
```

AVMusorin replied:
Fixed.

azat commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -268,69 +268,103 @@ namespace
-                    kafka_config.set(setting_name_in_kafka_config, setting_value);
+                    if (topic_name != topic)
+                        continue;
+                    loadTopicConfig(kafka_config, config, kafka_topic_path);
```
I was thinking that separate consumer/producer settings should support per-topic syntax as well, and both of them, this one and simpler.

AVMusorin replied:
Do you mean `<kafka_TOPIC_NAME>`? I thought that it is a part of the legacy... (regarding code comments)

Maybe it is better to reject `<kafka_TOPIC_NAME>` at all? Simpler code / less config variations.

azat replied:
>Do you mean <kafka_TOPIC_NAME>? I thought that it is a part of the legacy... (regarding code comments)

Yes, it is deprecated syntax, but maybe it still make sense to support it, since it is not a lot of code, but easier syntax (usually you don't have special symbols in the topic names and you are allowed to use easier syntax instead of more "advanced " one).

But I don't have strong opinion here, I'm OK to leave it as-is.

AVMusorin replied:
Moved the legacy topic parsing to a separate function (will be easier to deprecate in the future) and allowed to use it (old schema) inside `<consumer>/<producer>` tags.

azat commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -870,6 +899,26 @@ void StorageKafka::updateConfiguration(cppkafka::Configuration & kafka_config)
+    if (config.has(config_prefix))
+        loadConsumerConfig(kafka_config, config, config_prefix, topics);
+
+    loadTopicLegacyConfig(kafka_config, config, config_prefix, topics);
```
Should this be inside `loadConsumerConfig`? Since without it will not have `consumer` in the config prefix and will load nothing I guess (same for producer)

AVMusorin replied:
Yes, makes sense.

AVMusorin commented:
Could please someone else take a look at this PR?

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #58956 from AVMusorin/ClickHouse:kafka-settings -->
*This is an automated comment for commit 0cd0aaa6779a54dbc84ba0fea3209d3e8205b051 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58956/0cd0aaa6779a54dbc84ba0fea3209d3e8205b051/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58956/0cd0aaa6779a54dbc84ba0fea3209d3e8205b051/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

AVMusorin commented:
@rschu1ze Could you please take a look at this PR? Since I partially rewrote your code.

rschu1ze commented:
> @rschu1ze Could you please take a look at this PR? Since I partially rewrote your code.

Ah, that would be #58956, right? Yes, I can check in a bit.

rschu1ze commented:
If you don't mind, could you please explain the actual root cause of #58983? It is neither clear to me from the issue nor from the PR message.

AVMusorin commented:
> @AVMusorin If you don't mind, could you please explain the actual root cause of #58983? It is neither clear to me from the issue nor from the PR message.

Yes. I've also updated the description.

In short, `librdkafka` prints this kind of warning if consumer properties are used for producer instances or producer properties for consumers. https://github.com/ClickHouse/librdkafka/blob/master/src/rdkafka_conf.c#L3940

The main idea of this PR - is to allow to specify settings for a consumer and a producer in separate config sections (but under `kafka` tag). It is clearer what settings will be applied for a consumer and what for a producer, also, it prevents printing annoying messages in the log...

rschu1ze commented on tests/integration/test_storage_kafka/configs/kafka.xml:
```diff
@@ -11,16 +10,39 @@
             <!-- default: 3000 -->
             <heartbeat_interval_ms>300</heartbeat_interval_ms>
+        </kafka_topic>
+        <consumer>
```
Cosmetic: Let's use 4 spaces indentation instead of 2.

rschu1ze commented on tests/integration/test_storage_kafka/configs/kafka.xml:
```diff
@@ -11,16 +10,39 @@
+          <auto_offset_reset>earliest</auto_offset_reset>
+            <kafka_topic>
+            <!-- Setting for topic will be applied only for consumer -->
+            <name>consumer_hang</name>
```
Please intend this, it's visually hard to keep things apart.

AVMusorin replied:
Sorry for this. Fixed.

rschu1ze commented on tests/integration/test_storage_kafka/configs/kafka.xml:
```diff
@@ -11,16 +10,39 @@
+        </kafka_topic>
+        <consumer>
+          <auto_offset_reset>earliest</auto_offset_reset>
+            <kafka_topic>
```
Please don't intend this.

rschu1ze commented on tests/integration/test_storage_kafka/configs/kafka.xml:
```diff
@@ -11,16 +10,39 @@
+          </kafka_topic>
+        </consumer>
+        <producer>
+          <!-- NOTE: consumer_hang - name of the topic -->
```
L. 36 talks about topic `consumer_hang` but I don't see where the XML on l. 36 - 45 does anything specifically for `consumer_hang`.

AVMusorin replied:
Fixed

rschu1ze commented on tests/integration/test_storage_kafka/configs/kafka.xml:
```diff
@@ -11,16 +10,39 @@
+        </consumer>
+        <producer>
+          <!-- NOTE: consumer_hang - name of the topic -->
+          <kafka_separate_settings>
```
What is `kafka_separate_settings`? It does not look like a kafka topic (otherwise it would be wrapped in `<kafka_topic><name>`). Why do we need it? Why can't the producer settings simply be listed under `<producer>`?

AVMusorin replied:
`<kafka_separate_settings>` this is the old syntax I left it here to check that the warning works fine.

rschu1ze commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -846,6 +879,24 @@ void StorageKafka::updateConfiguration(cppkafka::Configuration & kafka_config)
+    const auto & config = getContext()->getConfigRef();
+    auto config_prefix = getConfigPrefix();
+    if (config.has(config_prefix))
+        loadProducerConfig(kafka_config, config, config_prefix, topics);
```
Silly question but why don't we `loadTopicLegacyConfig()` as in l. 889?

AVMusorin replied:
I agreed to remove this legacy syntax at all and throw a warning if it is used.

rschu1ze commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -244,69 +244,130 @@ namespace
+            return;
+        loadFromConfig(kafka_config, config, producer_path, topics);
+        /// Load legacy configuration per topic from the same producer level
+        loadTopicLegacyConfig(kafka_config, config, producer_path, topics);
```
I actually thought that we should remove the legacy syntax (basically `loadTopicLegacyConfig()`), now that it has been deprecated for a year (#46752). Perhaps that would simplify the code.

AVMusorin replied:
Yes, agree! I removed it and added a warning if someone still uses it.

rschu1ze commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -653,6 +710,19 @@ cppkafka::Configuration StorageKafka::getConsumerConfiguration(size_t consumer_n
+    conf.set("metadata.broker.list", brokers);
+    conf.set("client.id", client_id);
+    conf.set("client.software.name", VERSION_NAME);
+    conf.set("client.software.version", VERSION_DESCRIBE);
```
Cosmetic: + newline before l. 720, also in l. 702.

AVMusorin replied:
Done.

rschu1ze commented on src/Storages/Kafka/StorageKafka.cpp:
```diff
@@ -244,69 +244,130 @@ namespace
+        // The legacy syntax has the problem that periods in topic names (e.g. "sports.football") are not supported because the Poco
+        // configuration framework hierarchy is based on periods as level separators. Besides that, per-topic tags at the same level
+        // as <kafka> are ugly.
+        String topic_path = prefix;
```
The stuff on l. 341-343 is new and I don't get why it is needed.

AVMusorin replied:
Also, removed.

rschu1ze reviewed (commented):
I tried to verify the correctness of this PR by looking at the code but too many things were thrown from one place to another, renamed and slightly changed in meaning, so it is hard to be sure. I could install Kafka and verify by debugging but I remember the horrors of getting Kafka set up one year ago for my other PR.

Please have a look at my comments. I am happily to merge this PR if you think that everything is really okay.

From my view, we can drop the legacy syntax for Kafka topics (I wrote it in a comment). I would be happy if you like to do this in this PR (or in another one), I guess this would also simplify the code.

rschu1ze commented:
@antaljanosbenjamin Just FYI since you were working on Kafka recently

rschu1ze commented:
Kindly take a look at the remaining comments, then I can happily merge. Thanks!

AVMusorin commented:
> Kindly take a look at the remaining comments, then I can happily merge. Thanks!

Yes, sure. I'm in the process and let you know when it's done. Thanks!

alexey-milovidov commented:
@AVMusorin, How is the progress?

AVMusorin committed:
refactore: improve reading several configurations for kafka

Simplify and do some refactoring for kafka client settings.

Allows to set up separate
settings for consumer and producer like:

```
<consumer>
    ...
</consumer>

<producer>
    <kafka_topic>
        <name>topic_name</name>
        ...
    </kafka_topic>
</producer>
```

Moreover, this fixes warnings from kafka client like:
`Configuration property session.timeout.ms is a consumer property and
will be ignored by this producer instance`

AVMusorin commented on programs/server/config.xml:
```diff
@@ -1553,6 +1553,43 @@
+        <heartbeat_interval_ms>301</heartbeat_interval_ms>
+        <statistics_interval_ms>600</statistics_interval_ms>
+        -->
+        <!-- End Global configuration -->
```
Fixed.

AVMusorin commented on programs/server/config.xml:
```diff
@@ -1553,6 +1553,43 @@
         </tables>
     </rocksdb>
     -->
+    <!-- <kafka> -->
```
Yep. Done.

AVMusorin commented:
A short recap of what've done since the last changes:
* fixes review comments
* remove legacy syntax to apply kafka settings per topic (this helped to simplify code). The warning will be generated highlight that someone till uses the deprecated syntax.
* added trace logs with what exactly settings were read and applied for consumer and producer. Helps to debug especially since the last changes kafka settings can be read not only from CH config files but also from named collections.

AVMusorin commented on programs/server/config.xml:
```diff
@@ -1553,6 +1553,43 @@
+        <kafka_topic>
+            <name>football</name>
+            <request_timeout_ms>6000</request_timeout_ms>
+        </kafka_topic>
```
They should work for now since I haven't removed the processing of the old syntax.

rschu1ze commented on docs/en/engines/table-engines/integrations/kafka.md:
```diff
@@ -170,53 +170,37 @@ Similar to GraphiteMergeTree, the Kafka engine supports extended configuration u
-    </kafka_topic>
+    <statistics_interval_ms>3000</statistics_interval_ms>
+
+    <!-- Settings for consumer -->
```
Stupid question but is it no longer possible to have `<kafka_topic>` defined directly under `<kafka>`? Asking because the PR description says that is possible.

AVMusorin replied:
Yes, it is possible to define `<kafka_topic>` directly under `<kafka>`. I'll add this case in the documentation.

rschu1ze reviewed (approved):
Thanks.

The codebase is still in a stabilization phase, I'll hopefully be able to merge this next week.

AVMusorin committed:
docs: added kafka_topic directly under kafka tag
